File: src/com/leetcode/tree/Basic_Queue.java
package com.leetcode.tree;

import com.leetcode.linkedlist.ListNode;
import java.util.Arrays;


/**
 * Created by Michael on 2017/3/9.
 * The basic implementation of <Queue> Data Structure.
 * 队列的两种实现方式：链表和数组。
 *
 * <Tags>
 * - Implement Queue using Linkedlist.
 * - Implement Queue using Array.
 * - Two Pointers: head and tail [ head -> ... -> ... -> tail -> ]
 *
 */
public class Basic_Queue {

    /** 队列的两种实现媒介：数组和链表 */
    // 方式1：使用数组实现（循环数组）：利用数组的索引值记录队首和队尾的位置。需要靠模运算区分队列是满还是空。
    // 方式2：使用链表实现（循环链表）：利用节点相邻关系判断队列满空。

    /** 队首和队尾指针的定义 */
    // 队首指针 head：指向队首元素。
    // 队尾指针 tail：指向队尾元素的下一个元素（而不是队尾元素本身）。

    /** 实现队列时的核心问题：如何判断队满和队空 */
    // 如果不加任何判断方式，队列满和队列空的表现都是队首指针和队尾指针重合。
    // 有两种方式来区分队满和队空：
    // 方式1：将队列<使用长度>限制为<实际容量>减1，
    //       对于链表，队列空时head == tail，队列满时tail.next == head
    //       对于数组，队列空时head == tail，队列满时(tail + 1) % length == head
    // 方式2：设置一个flag指示队列当前是否为空。队列空和队列满的时候都有head == tail，但是可以通过flag状态区分是队列空还是队列满。
    // 通常使用方式1。

    /** 队列基本操作的时间复杂度 */
    //  1.入队：o(1)
    //  2.出队：o(1)
    //  3.队首：o(1)

    //  Designed Capacity = 5
    //  True Capacity = 6
    //  0 -> 0 -> 0 -> 0 -> 0 -> 0 -> (loop)  head == tail 队列空
    //  ↑
    //  head,tail
    //
    //  1 -> 2 -> 3 -> 4 -> 5 -> 0 -> (loop)  tail.next == head 队列满
    //  ↑                        ↑
    //  head                     tail
    //
    //  1 -> 2 -> 3 -> 4 -> 5 -> 0 -> (loop)  head == tail 队列空
    //                           ↑
    //                           head,tail
    //
    //  7 -> 8 -> 9 -> 10 -> 0 -> 6 -> (loop)  tail.next == head 队列满
    //                       ↑    ↑
    //                       tail head
    public static void main(String[] args) {
        Queue_LinkedList q = new Queue_LinkedList();
        q.add(1);
        q.add(2);
        q.add(3);
        q.add(4);
        q.add(5);
        q.add(6);
        q.remove();
        q.remove();
        q.remove();
        q.remove();
        q.remove();
        q.peek();
        q.add(6);
        q.add(7);
        q.add(8);
        q.add(9);
        q.add(10);
        System.out.println(q.peek());

        Queue_Array q2 = new Queue_Array();
        q2.add(1);
        q2.add(2);
        q2.add(3);
        q2.add(4);
        q2.add(5);
        q2.add(6);
        q2.remove();
        q2.remove();
        q2.remove();
        q2.remove();
        q2.remove();
        System.out.println(q2.peek());
        q2.add(6);
        q2.add(7);
        q2.add(8);
        q2.add(9);
        q2.add(10);
        q2.add(6);
        System.out.println(q2.peek());
    }
}

/** 使用<链表>实现队列 */
// 判断队列满的条件：tail.next == head
// 判断队列空的条件：head == tail
class Queue_LinkedList {
    private ListNode queue;
    private ListNode head;
    private ListNode tail;
    private ListNode start;
    private final int DEFAULT_LEN = 5;

    // Default constructor with default queue size.
    public Queue_LinkedList() {
        queue = ListNode.Generator(new int[DEFAULT_LEN + 1]);
        start = queue;
        head = queue;
        tail = queue;
        while (queue.next != null) queue = queue.next;      // Create loop
        queue.next = head;
    }

    public Queue_LinkedList(int capacity) {
        queue = ListNode.Generator(new int[capacity + 1]);
        head = queue;
        tail = queue;
        while (queue.next != null) queue = queue.next;      // Create loop
        queue.next = head;
    }

    public boolean add(int x) {
        if (tail.next == head) {
            System.out.println("Error: Queue is full.");
            return false;        // 队列满时，禁止再加入元素
        }
        tail.val = x;
        tail = tail.next;
        start.print();
        return true;
    }

    public int remove() {
        if (head == tail) {
            System.out.println("Error: Queue is empty.");
            return -1;                // 队列空时，禁止再删除元素
        }
        int ret = head.val;
        head.val = 0;
        head = head.next;
        start.print();
        return ret;
    }

    public int peek() {
        if (head == tail) {
            System.out.println("Error: Queue is empty.");
            return -1;                // 队列空时，禁止返回元素
        }
        start.print();
        return head.val;
    }
}

/** 使用<数组>实现队列 */
// 判断队列满的条件：(tail + 1) % len == head 其含义是：tail的下个元素就是head（这里“下个元素”的概念是循环长度的下个元素，因此需要使用模运算实现）
// 判断队列空的条件：head == tail

// 示例：
// 队列满的情况1
// [1, 2, 3, 4, 5, 0]
//  ↑              ↑
//  head           tail    tail的循环下个元素就是head

// 队列满的情况2
// [4, 5, 0, 1, 2, 3]
//        ↑  ↑
//      tail head          tail的下个元素是head
class Queue_Array {
    int[] queue;
    private int head;
    private int tail;
    private final int DEFAULT_LEN = 5;

    public Queue_Array() {
        queue = new int[DEFAULT_LEN + 1];
        head = 0;
        tail = 0;
        System.out.println(Arrays.toString(queue));
    }

    public boolean add(int x) {
        int l = queue.length;
        if ((tail + 1) % l == head) {                       // 判断队列满的条件
            System.out.println("Error: Queue is full.");
            return false;
        }
        queue[tail] = x;
        tail = (tail + 1) % l;
        System.out.println(Arrays.toString(queue));
        return true;
    }

    public int remove() {
        if (head == tail) {                                 // 判断队列空的条件
            System.out.println("Error: Queue is empty.");
            return -1;
        }
        int ret = queue[head];
        queue[head] = 0;
        head = (head + 1) % queue.length;
        System.out.println(Arrays.toString(queue));
        return ret;
    }

    public int peek() {
        if (head == tail) {                                  // 判断队列空的条件
            System.out.println("Error: Queue is empty.");
            return -1;
        }
        return queue[head];
    }

}

File: src/com/leetcode/tree/Basic_Queue.cc
#include "../common.h"
#include <vector>
#include <list>
#include <algorithm>
using namespace std;

/**
 * Created by Michael on 2019/8/10.
 * 
 * 循环队列 Circular Queue 使用数组或循环链表实现。
 * 基本功能：add，remove，peek
 * 队列vs栈：先进先出vs先进后出。
 * 
 * <循环队列的实现>
 * 0. 内存申请
 *      - 与栈边用边申请边扩充的方式不同，循环队列具有固定的容量大小
 *      - 如果使用数组，正常申请即可。如果不使用 length，则申请量必须等于需求量 + 1
 *      - 如果使用链表，则需要初始化好整个链表的节点，并且使其成环。如果不使用 length，则节点数量必须等于需求量 + 1
 * 
 * 1. 队首指针与队尾指针的规则：
 *      - 队首 head 指向 <队首元素> 
 *      - 队尾 tail 指向 <队尾元素的下一个有效位置>（并不是队尾元素本身），tail 可能小于 head
 *      - 添加元素：先将数据写到 tail 所在位置，再右移 tail，最后修正 tail 至有效取值范围
 *      - 删除元素：先将数据从 head 所在位置导出，再右移 head，最后修正 head 至有效取值范围
 * 
 * 2. 队列判空与队列判满的规则：
 *      - 如果使用 length 记录队列的有效元素个数，那么判空和判满只需要看 length 的值即可，非常简单
 *      - 如果非不用 length，则需要空出一个位置，如下操作：
 *      - 如果申请内存空间的长度为 n，则队列最多只能放 n - 1 个元素。这么做是为了区分<队列空>和<队列满>这两种状态。
 *      - 队列判空：head == tail
 *      - 队列判满：
 *          - 数组实现：(tail + 1) % capacity == head
 *          - 链表实现：tail->next == head
 * 
 */

class ArrayQueue {
private:
  vector<int> data;
  size_t head = 0;
  size_t tail = 0;
  size_t length = 0;

public:
  ArrayQueue(size_t capacity) {
    data.resize(capacity);
  }

  bool enqueue(int x) {
    if (length == data.size()) return false;
    data[tail++] = x;       // 先赋值，再扩张指针
    tail %= data.size();    // 修正指针的位置，使其返回 0 - (capacity-1) 取值区间
    length++;
    return true;
  }

  // 入队指定长度的数据，返回成功入队的数据长度
  size_t enqueue(int* seq, size_t len) {
    size_t valid = min(data.size() - length, len);
    for (int i = 0; i < valid; i++) {
      data[tail++] = seq[i];    // 先赋值，后扩张指针
      tail %= data.size();      // 修正指针位置
    }
    length += valid;
    return valid;
  }

  int dequeue() {
    if (length == 0) return -1;
    int result = data[head++];    // 先缓存，后收缩指针
    head %= data.size();          // 同样需要修正指针位置
    length--;
    return result;
  }

  // 出队指定长度的数据，返回成功出队的数据长度
  size_t dequeue(int* seq, size_t len) {
    size_t valid = min(length, len);
    for (int i = 0; i < valid; i++) {
      seq[i] = data[head++];      // 先导出数组，后收缩指针
      head %= data.size();        // 最后修正指针位置
    }
    length -= valid;
    return valid;
  }
};



class CircularQueueLL {

private:
  ListNode* head = nullptr;
  ListNode* tail = nullptr;

public:
  // 需求量为 capacity，但需要创建 + 1 个节点
  CircularQueueLL(int capacity) {
    head = new ListNode(0);
    tail = head;
    ListNode* curr = head;
    for (int i = 0; i < capacity; i++) {
      curr->next = new ListNode(0);
      curr = curr->next;
    }
    curr->next = head;    // Make it loop
  }

  // Add element to the tail of this queue.
  ListNode* enqueue(int val) {
    if (tail->next == head) {             // Full Check
      std::cout << "Queue is full. Cannot add element." << std::endl;
        return nullptr;
    }
    tail->val = val;                      // Update val.
    tail = tail->next;                    // Move tail to new tail.
  }

  // Remove element from the head of this queue.
  int dequeue() {
    if (head == tail) {                 // Empty Check
      std::cout << "Queue is empty. Cannot remove element." << std::endl;
      return -1;
    }
    int res = head->val;
    head->val = 0;                       // Clean val.
    head = head->next;                   // Move head to new head.
    return res;
  }

  int peek() {
    // Check if empty first
    if (head == tail) {
      std::cout << "Queue is empty. Cannot remove element." << std::endl;
      return -1;
    }
    return head->val;
  }
};

int main() {
  
  ArrayQueue* aq = new ArrayQueue(5);
  aq->enqueue(99);
  aq->enqueue(88);
  aq->dequeue();
  vector<int> d = {1, 2, 3, 4};
  aq->enqueue(d.data(), d.size());
  vector<int> dd(5);
  aq->dequeue(dd.data(), 10);
  print(dd);


  CircularQueueLL* q = new CircularQueueLL(3);
  q->enqueue(2);
  q->enqueue(5);
  q->enqueue(7);
  q->enqueue(10);
  q->dequeue();
  q->dequeue();
  std::cout << q->peek() << std::endl;
  q->dequeue();
  q->dequeue();
}

/* 数组实现循环队列示意图：队尾与队首的位置不确定，但是可以通过求模解决 (tail + 1) % length == head
  Capacity = 3, Actual Size = 4
  ------------------------------------------------------------------------------------------
  head                  tail
   ↓                     ↓           (3 + 1) % 4 == 0  -> full
   0       1       2     3
   22     33      44     0
  ------------------------------------------------------------------------------------------
          tail   head
           ↓      ↓                  (1 + 1) % 4 == 2 -> full
    0      1      2      3
   67      0     55     84
  ------------------------------------------------------------------------------------------

 */